src/concurrency/thread.rs

```diff
@@ -620,11 +630,26 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
     /// long as we can and switch only when we have to (the active thread was
     /// blocked, terminated, or has explicitly asked to be preempted).
     fn schedule(&mut self, clock: &Clock) -> InterpResult<'tcx, SchedulingAction> {
-        // Check whether the thread has **just** terminated (`check_terminated`
-        // checks whether the thread has popped all its stack and if yes, sets
-        // the thread state to terminated).
-        if self.threads[self.active_thread].check_terminated() {
-            return Ok(SchedulingAction::ExecuteDtors);
+        // If we are the main thread, and our call stack is empty but our state is Enabled, we are
+        // about to terminate.
+        // But, if there are any other threads which can execute, yield to them instead of falling
+        // through to the termination state.
+        let active_thread = &self.threads[self.active_thread];
+        if self.active_thread == MAIN_THREAD
+            && active_thread.stack.is_empty()
+            && active_thread.state == ThreadState::Enabled
+            && self.threads.iter().any(|t| t.state == ThreadState::Enabled && !t.stack.is_empty())
+            && self.main_thread_yields_at_shutdown_remaining > 0
+        {
+            self.yield_active_thread = true;
+            self.main_thread_yields_at_shutdown_remaining -= 1;
```

Comment:
Ideally we'd do this yielding *after* executing main thread TLS dtors (because it'd be UB to access them). Not sure how terrible that would be to implement though, so a FIXME is also fine.

Reply:
Yeah... I would like to understand how to do this correctly. The naive implementation is just shifting around the big check this PR adds so that the original `if should_terminate() { return Ok(ExecuteDtors) }` comes first, but that seems to produce data races whenever this yielding behavior kicks in: We run the main thread's dtors, then schedule another thread on the next step, and that thread reports a data race on one of a few allocations in the Rust runtime.

It wouldn't surprise me if these data races are legitimate. After all, we do have this comment:
```
                // The main thread terminated; stop the program.
                // We do *not* run TLS dtors of remaining threads, which seems to match rustc behavior.
 ```

Reply:
Hard to say more without knowing the details of the race and how you implemented this. But we already currently will automatically yield to other threads while main thread Dtors are running, so we should already be seeing such races?

The best place to put this yielding behavior seems to be in the `if self.threads[MAIN_THREAD].state == ThreadState::Terminated` check, no? Instead of stopping, we keep going, but with some kind of bound.

Reply:
Actually a bound on the steps is probably better than yielding... if automatic yielding is disabled, this might otherwise keep going forever even after the main thread stopped.

Reply:
I pushed a commit which does the change I was trying to describe above. I've seen two data races reported, one in the test suite and the other when I run this file:
```rust
use std::thread;

fn main() {
    thread::scope(|s| {
        s.spawn(|| {});
    });
}
```
With `./miri many-seeds ./miri demo.rs`

I agree that a bound on the extra steps sounds reasonable. Is the concern that this yielding behavior may take programs which currently exit under the non-yielding option and turn them into programs that never exit?

Reply:
>  Is the concern that this yielding behavior may take programs which currently exit under the non-yielding option and turn them into programs that never exit?

Yes.

Reply:
The TLS shims re-enable terminated threads. Is that a good thing? Do they have to be Enabled in order to execute user code?

Reply:
Yes, threads must definitely be enabled when running TLS.

I wish they wouldn't be disabled and then re-enabled... this entire scheduler and TLS stuff needs a rework at some point, but I haven't yet found a nice way to factor this. I still think a generator would be good... basically `yield`ing a function to run, which is then executed until the stack is empty again, at which point the generator is consulted to figure out whether there is another function to run or whether this thread is done.

Reply:
Ah sadly using generators for this [will not work](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Generators.20that.20borrow.20on.20each.20resume.3F).